source/common/common/callback_impl.h

```diff
@@ -81,5 +105,82 @@ template <typename... CallbackArgs> class CallbackManager {
   const std::shared_ptr<bool> still_alive_{std::make_shared<bool>(true)};
 };
 
+/**
+ * @brief Utility class for managing callbacks across multiple threads.
+ *
+ * Callback registration (via locks) and callback execution (via dispatchers and shared_ptr's) are
+ * thread-safe. However the main ThreadSafeCallbackManager instance is assumed to exist on a single
+ * thread during it's lifetime, the same as the provided dispatcher it is constructed with.
+ *
+ * @note This is not safe to use for instances in which the lifetimes of the threads registering
+ * callbacks is less than the thread that owns the callback manager due to an assumption that
+ * dispatchers registered alongside callbacks remain valid, even if the callback expires.
+ *
+ * @see CallbackManager for a non-thread-safe version
+ */
+class ThreadSafeCallbackManager {
+  struct CallbackHolder;
+  using CallbackListEntry = std::tuple<std::weak_ptr<CallbackHolder>, Event::Dispatcher&>;
+
+public:
+  using Callback = std::function<void()>;
+
+  /**
+   * @param dispatcher Dispatcher relevant to the thread in which the callback manager is
+   * created/managed
+   */
+  explicit ThreadSafeCallbackManager(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+  /**
+   * @brief Add a callback.
+   * @param dispatcher Dispatcher from the same thread as the registered callback. This will be used
+   *                   to schedule the execution of the callback.
+   * @param callback callback to add
+   * @return ThreadSafeCallbackHandlePtr a handle that can be used to remove the callback.
+   */
+  ABSL_MUST_USE_RESULT ThreadSafeCallbackHandlePtr add(Event::Dispatcher& dispatcher,
+                                                       Callback callback);
```

Comment:
Going back to the code in the followup PR: https://github.com/envoyproxy/envoy/pull/16201/files#r650306747

I wonder if there's some potential for child callbacks to not need to use weak_ptr themselves to tell if it is safe to invoke the main body of the callback.  Here's an idea:
- Require that ThreadSafeCallbackHandlePtr be deleted in the dispatcher thread by calling `ASSERT(dispatcher.isThreadSafe());` from the ThreadSafeCallbackHandlePtr destructor
- Since ThreadSafeCallbackHandlePtr destruction and the callback both execute in the dispatcher thread, you can tell if the callback is still usable if a still_alive owned by the ThreadSafeCallbackHandlePtr is still alive.

I think this would also allow you to use unique_ptr instead of shared_ptr for ThreadSafeCallbackHandlePtr.  Also, I think that would allow you to return unique_ptr from DrainManagerImpl::createChildManager

Reply:
I think this approach would lead to a race-condition on callback execution. In particular I would now see the following order of events being possible:

0. The `std::unique_ptr<CallbackHolder>` would begin destruction. Asserting it's running on the dispatcher thread, then attempting to take a lock with the parent to remove the callback. 
1. The lock is already held because the manager is scheduling all callbacks to run, posting to dispatchers.
2. The lock is released and taken up by the destructor. The callback is removed from the `callbacks_` list
3. The dispatcher runs the block of code scheduled by the manager to run the callback. The now-free'd callback is attempted to be executed.

Because dispatcher's async nature, I'm not sure there is a way to get away from using `weak_ptr` to validate the callback is still valid before executing.

Reply:
I don't know why your reply to this comment is not showing up here.  Quoting below:

> I think this approach would lead to a race-condition on callback execution. In particular I would now see the following order of events being possible:
>
> 0. The std::unique_ptr<CallbackHolder> would begin destruction. Asserting it's running on the dispatcher thread, then attempting to take a lock with the parent to remove the callback.
> 1. The lock is already held because the manager is scheduling all callbacks to run, posting to dispatchers.
> 2. The lock is released and taken up by the destructor. The callback is removed from the callbacks_ list
> 3. The dispatcher runs the block of code scheduled by the manager to run the callback. The now-free'd callback is attempted to be executed.
>
> Because dispatcher's async nature, I'm not sure there is a way to get away from using weak_ptr to validate the callback is still valid before executing.

The callback proper is copied as it is posted to the dispatcher.  On initial construction of the callback, you'ld take a weap_ptr<bool> to CallbackHolder::still_alive_ as part of the callback capture.  On execution, you'ld return if `still_alive.expired()`

Reply:
Good call. I have removed the use of `shared_ptr` and moved the `weak_ptr` usages to a `still_alive_` value. Now the callback handles and the drain-manager children are contained with `unique_ptr`s.